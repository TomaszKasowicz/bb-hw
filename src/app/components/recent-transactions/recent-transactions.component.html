<app-window-header [icon]="'./assets/icons/briefcase.png'">
  Recent Transactions
</app-window-header>
<div class="flex user-menu">
  <app-search (searchTextChanged)="setFilter($event)"></app-search>
  <app-sort (sortChange)="setSorting($event)"></app-sort>
</div>
<div class="transaction-list">
  <app-transaction-details class="flex" *ngFor="let transaction of transactions$ | async as list" [transaction]="transaction"
  [attr.data-testid]="'merchant-' + transaction.merchant"></app-transaction-details>
</div>
