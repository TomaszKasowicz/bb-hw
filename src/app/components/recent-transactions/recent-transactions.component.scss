:host {
  display: flex;
  flex-direction: column;
}

.user-menu {
  justify-content: space-between;
  padding: 1em 2em;
  gap: 1em;
  & > * {
    flex-grow: 1;
  }
}

.transaction-list {
  padding: 0em 1em;
  overflow: auto;
}
