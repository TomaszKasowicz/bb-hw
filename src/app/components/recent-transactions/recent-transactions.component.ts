import { Component, OnInit } from '@angular/core';
import { BehaviorSubject, merge, Observable } from 'rxjs';
import { map, switchMapTo, withLatestFrom } from 'rxjs/operators';
import { SortState } from '../../model/sort-state';
import { Transaction } from '../../model/transaction';
import { TransactionsService } from '../../services/transactions.service';

@Component({
  selector: 'app-recent-transactions',
  templateUrl: './recent-transactions.component.html',
  styleUrls: ['./recent-transactions.component.scss']
})
export class RecentTransactionsComponent implements OnInit {
  public transactions$: Observable<Transaction[]>;

  private filter$ = new BehaviorSubject('');
  private sort$ = new BehaviorSubject<SortState>({
    property: 'transactionDate',
    ascending: false
  });

  constructor(private transactionsService: TransactionsService) {}

  ngOnInit(): void {
    const rawTransactions$ = this.transactionsService.getRecentTransactions();
    this.transactions$ = merge(rawTransactions$, this.filter$, this.sort$).pipe(
      switchMapTo(rawTransactions$),
      withLatestFrom(this.filter$),
      map(([transactions, filterText]) =>
        this.filterTransactions(transactions, filterText)
      ),
      withLatestFrom(this.sort$),
      map(
        ([filteredTransactions, sort]: [
          fiteredTranactions: Transaction[],
          sort: SortState
        ]) => this.sortTransactions(filteredTransactions, sort)
      )
    );
  }

  setFilter(text: string) {
    this.filter$.next(text);
  }

  setSorting(sort: SortState) {
    this.sort$.next(sort);
  }

  private filterTransactions(
    transactions: Transaction[],
    filterText?: string
  ): Transaction[] {
    if (!filterText) {
      return transactions;
    }

    const caseInsensitiveFilter = filterText.toLowerCase();
    return transactions.filter(
      t =>
        t.merchant.toLowerCase().includes(caseInsensitiveFilter) ||
        t.transactionType.toLowerCase().includes(filterText)
    );
  }

  private sortTransactions(
    transactions: Transaction[],
    sort: SortState
  ): Transaction[] {
    const { property, ascending } = sort;
    return transactions.sort((a: Transaction, b: Transaction) => {
      const aValue: string | number | undefined =
        a[property as keyof Transaction];
      const bValue: string | number | undefined =
        b[property as keyof Transaction];

      if (typeof aValue === 'number' && typeof bValue === 'number') {
        return ascending ? aValue - bValue : bValue - aValue;
      }

      return ascending
        ? (aValue as string).localeCompare(bValue as string)
        : (bValue as string).localeCompare(aValue as string);
    });
  }
}
