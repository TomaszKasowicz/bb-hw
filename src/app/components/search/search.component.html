<input class="flex-grow-1" type="text" [(ngModel)]="searchText" placeholder="Search By Typing" (keyup)="emitChange($event)">
<button [class.hidden]="!searchText" (click)="clear()">X</button>