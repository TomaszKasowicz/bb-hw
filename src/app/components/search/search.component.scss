:host {
  display: flex;
}

input {
  padding-right: 10px;
}

button {
  margin-left: -5px;
  background-color: var(--color-background-primary);
  border: none;
  border-bottom: 1px solid var(--color-line);
  color: var(--color-text-postfix);

  &.hidden {
    visibility: hidden;
  }
  &:hover {
    cursor: pointer;
  }
}
