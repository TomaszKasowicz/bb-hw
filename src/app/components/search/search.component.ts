import { Component, EventEmitter, OnInit, Output } from '@angular/core';

@Component({
  selector: 'app-search',
  templateUrl: './search.component.html',
  styleUrls: ['./search.component.scss']
})
export class SearchComponent implements OnInit {
  @Output()
  public searchTextChanged = new EventEmitter<string>();
  public searchText: string;

  constructor() {}

  ngOnInit(): void {}

  emitChange(event?: KeyboardEvent) {
    if (event?.key === 'Escape') {
      this.searchText = '';
    }
    this.searchTextChanged.emit(this.searchText);
  }

  clear() {
    this.searchText = '';
    this.emitChange();
  }
}
