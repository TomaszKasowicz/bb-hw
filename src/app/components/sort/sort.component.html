<label>
  Sort By:
</label>
<button class="flex-grow-1" *ngFor="let sortColumn of sortColumns"
(click)="toggleSort(sortColumn.propertyName)"
>
  <div class="flex">
    <label>{{ sortColumn.label }}</label>    
    <div class="arrow" 
    [class.arrow-hidden]="sortColumn.propertyName !== state.property"
    [class.arrow-up]="state.ascending" 
    [class.arrow-down]="!state.ascending"></div>  
  </div>
</button>