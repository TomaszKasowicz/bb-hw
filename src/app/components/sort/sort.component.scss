:host {
  display: flex;
}

label {
  margin-right: 1em;
}

button {
  background-color: var(--color-background-primary);
  border: 1px solid var(--color-line);

  & div {
    justify-content: center;
  }
}

.arrow {
  margin-left: 0.5em;
  width: 0;
  height: 0;
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  flex-grow: 0;

  &-hidden {
    visibility: hidden;
  }

  &-up {
    border-bottom: 5px solid #999999;
  }

  &-down {
    border-top: 5px solid #999999;
  }
}
