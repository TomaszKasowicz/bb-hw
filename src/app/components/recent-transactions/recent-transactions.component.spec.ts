import {
  createComponentFactory,
  mockProvider,
  Spectator,
  SpyObject
} from '@ngneat/spectator/jest';
import { MockComponent } from 'ng-mocks';
import { of } from 'rxjs';
import { mockData } from '../../../mock/transactions';
import { TransactionsService } from '../../services/transactions.service';
import { SearchComponent } from '../search/search.component';
import { SortComponent } from '../sort/sort.component';
import { WindowHeaderComponent } from '../window-header/window-header.component';
import { RecentTransactionsComponent } from './recent-transactions.component';
import { TransactionDetailsComponent } from '../transaction-details/transaction-details.component';

describe('RecentTransactionsComponent', () => {
  let spectator: Spectator<RecentTransactionsComponent>;
  let service: SpyObject<TransactionsService>;

  const createComponent = createComponentFactory({
    component: RecentTransactionsComponent,
    declarations: [
      MockComponent(WindowHeaderComponent),
      MockComponent(SearchComponent),
      MockComponent(SortComponent),
      MockComponent(TransactionDetailsComponent)
    ],
    providers: [mockProvider(TransactionsService)],
    detectChanges: false
  });

  beforeEach(() => {
    spectator = createComponent();
    service = spectator.inject(TransactionsService);
  });

  it('should create', () => {
    expect(spectator.component).toBeTruthy();
  });

  it('should render empty list of transactions', () => {
    service.getRecentTransactions.mockReturnValue(of([]));
    spectator.detectChanges();

    expect(spectator.fixture).toMatchSnapshot();
    const detailElements = spectator.queryAll(TransactionDetailsComponent);
    expect(detailElements.length).toEqual(0);
  });

  describe('Transaction Filtering', () => {
    const transactions = mockData;

    it('Should render all transactions when filter is empty', () => {
      service.getRecentTransactions.mockReturnValue(of(transactions));
      spectator.detectChanges();
      expect(spectator.fixture).toMatchSnapshot();
      const detailElements = spectator.queryAll(TransactionDetailsComponent);
      expect(detailElements.length).toEqual(transactions.length);
    });

    it('Should render filtred transactions', () => {
      service.getRecentTransactions.mockReturnValue(of(transactions));
      spectator.detectChanges();
      spectator.triggerEventHandler(
        SearchComponent,
        'searchTextChanged',
        'tea'
      );
      spectator.detectChanges();

      expect(spectator.fixture).toMatchSnapshot();
      const detailElements = spectator.queryAll(TransactionDetailsComponent);
      expect(detailElements.length).toEqual(2);
    });
  });
});
