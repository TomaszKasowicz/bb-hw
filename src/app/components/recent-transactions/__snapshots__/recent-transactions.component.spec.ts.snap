// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentTransactionsComponent Transaction Filtering Should render all transactions when filter is empty 1`] = `
<app-recent-transactions
  __ngContext__={[Function LRootView]}
  filter$={[Function BehaviorSubject]}
  sort$={[Function BehaviorSubject]}
  transactions$={[Function Observable]}
  transactionsService={[Function Object]}
>
  <app-window-header>
     Recent Transactions
  
  </app-window-header><div
    class="flex user-menu"
  >
    <app-search />
    <app-sort />
  </div><div
    class="transaction-list"
  >
    <app-transaction-details
      class="flex"
      data-testid="merchant-The Tea Lounge"
    />
    <app-transaction-details
      class="flex"
      data-testid="merchant-Texaco"
    />
    <app-transaction-details
      class="flex"
      data-testid="merchant-The Tea Lounge"
    />
    <app-transaction-details
      class="flex"
      data-testid="merchant-Amazon Online Store"
    />
    <app-transaction-details
      class="flex"
      data-testid="merchant-7-Eleven"
    />
    <app-transaction-details
      class="flex"
      data-testid="merchant-H&M Online Store"
    />
    <app-transaction-details
      class="flex"
      data-testid="merchant-Jerry Hildreth"
    />
    <app-transaction-details
      class="flex"
      data-testid="merchant-Lawrence Pearson"
    />
    <app-transaction-details
      class="flex"
      data-testid="merchant-Whole Foods"
    />
    <app-transaction-details
      class="flex"
      data-testid="merchant-Southern Electric Company"
    />
    
  </div>
</app-recent-transactions>
`;

exports[`RecentTransactionsComponent Transaction Filtering Should render filtred transactions 1`] = `
<app-recent-transactions
  __ngContext__={[Function LRootView]}
  filter$={[Function BehaviorSubject]}
  sort$={[Function BehaviorSubject]}
  transactions$={[Function Observable]}
  transactionsService={[Function Object]}
>
  <app-window-header>
     Recent Transactions
  
  </app-window-header><div
    class="flex user-menu"
  >
    <app-search />
    <app-sort />
  </div><div
    class="transaction-list"
  >
    <app-transaction-details
      class="flex"
      data-testid="merchant-The Tea Lounge"
    />
    <app-transaction-details
      class="flex"
      data-testid="merchant-The Tea Lounge"
    />
    
  </div>
</app-recent-transactions>
`;

exports[`RecentTransactionsComponent should render empty list of transactions 1`] = `
<app-recent-transactions
  __ngContext__={[Function LRootView]}
  filter$={[Function BehaviorSubject]}
  sort$={[Function BehaviorSubject]}
  transactions$={[Function Observable]}
  transactionsService={[Function Object]}
>
  <app-window-header>
     Recent Transactions
  
  </app-window-header><div
    class="flex user-menu"
  >
    <app-search />
    <app-sort />
  </div><div
    class="transaction-list"
  >
    
  </div>
</app-recent-transactions>
`;
